### ISSUE TITLE
OptimizeForArchitecture.cmake skylake is detected as merom

### ISSUE THREAD

**Issue by baslr** [arangodb/arangodb#2035]
OptimizeForArchitecture.cmake  doesn't detect current Intel CPU gen.

```
devel branch

CMake Warning at cmake/OptimizeForArchitecture.cmake:120 (message):
  Your CPU (family 6, model 94) is not known.  Auto-detection of optimization
  flags failed and will use the 65nm Core 2 CPU settings.
Call Stack (most recent call first):
  cmake/OptimizeForArchitecture.cmake:169 (AutodetectHostArchitecture)
  CMakeLists.txt:384 (OptimizeForArchitecture)
```

using OptimizeForArchitecture.cmake and following from https://github.com/VcDevel/Vc results into

``````
arangodb-build/build$ cmake -DUSE_ROCKSDB=OFF -DCMAKE_BUILD_TYPE=Release ..
-- The CXX compiler identification is GNU 6.2.0
-- The C compiler identification is GNU 6.2.0
-- The ASM compiler identification is GNU
-- Found assembler: /usr/bin/cc
-- Check for working CXX compiler: /usr/bin/c++
-- Check for working CXX compiler: /usr/bin/c++ -- works
-- Detecting CXX compiler ABI info
-- Detecting CXX compiler ABI info - done
-- Detecting CXX compile features
-- Detecting CXX compile features - done
-- Check for working C compiler: /usr/bin/cc
-- Check for working C compiler: /usr/bin/cc -- works
-- Detecting C compiler ABI info
-- Detecting C compiler ABI info - done
-- Detecting C compile features
-- Detecting C compile features - done
-- Found PythonInterp: /usr/bin/python2 (found suitable exact version "2.7.12")
-- Performing Test _HAS_CXX11_FLAG
-- Performing Test _HAS_CXX11_FLAG - Success
-- Checking C++11 support for "__func__"
-- Checking C++11 support for "__func__": works
-- Checking C++11 support for "auto"
-- Checking C++11 support for "auto": works
-- Checking C++11 support for "auto_ret_type"
-- Checking C++11 support for "auto_ret_type": works
-- Checking C++11 support for "class_override_final"
-- Checking C++11 support for "class_override_final": works
-- Checking C++11 support for "constexpr"
-- Checking C++11 support for "constexpr": works
-- Checking C++11 support for "conditional"
-- Checking C++11 support for "conditional": works
-- Checking C++11 support for "decltype"
-- Checking C++11 support for "decltype": works
-- Checking C++11 support for "initializer_list"
-- Checking C++11 support for "initializer_list": works
-- Checking C++11 support for "lambda"
-- Checking C++11 support for "lambda": works
-- Checking C++11 support for "range_based_for_loop"
-- Checking C++11 support for "range_based_for_loop": works
-- Checking C++11 support for "nullptr"
-- Checking C++11 support for "nullptr": works
-- Checking C++11 support for "tuple"
-- Checking C++11 support for "tuple": works
-- Checking C++11 support for "regex"
-- Checking C++11 support for "regex": works
-- Checking C++11 support for "rvalue-references"
-- Checking C++11 support for "rvalue-references": works
-- Checking C++11 support for "sizeof_member"
-- Checking C++11 support for "sizeof_member": works
-- Checking C++11 support for "static_assert"
-- Checking C++11 support for "static_assert": works
-- Checking C++11 support for "variadic_templates"
-- Checking C++11 support for "variadic_templates": works
-- Checking C++11 support for "shared_ptr"
-- Checking C++11 support for "shared_ptr": works
-- Checking C++11 support for "unique_ptr"
-- Checking C++11 support for "unique_ptr": works
-- Checking C++11 support for "weak_ptr"
-- Checking C++11 support for "weak_ptr": works
-- Checking C++11 support for "thread"
-- Checking C++11 support for "thread": works
-- Checking C++11 support for "mutex"
-- Checking C++11 support for "mutex": works
-- Checking C++11 support for "noexcept"
-- Checking C++11 support for "noexcept": works
-- Looking for pthread.h
-- Looking for pthread.h - found
-- Looking for pthread_create
-- Looking for pthread_create - not found
-- Looking for pthread_create in pthreads
-- Looking for pthread_create in pthreads - not found
-- Looking for pthread_create in pthread
-- Looking for pthread_create in pthread - found
-- Found Threads: TRUE
-- target changed from "" to "auto"
-- Detected CPU: skylake
-- Performing Test check_cxx_compiler_flag__march_skylake
-- Performing Test check_cxx_compiler_flag__march_skylake - Success
-- Performing Test check_cxx_compiler_flag__msse2
-- Performing Test check_cxx_compiler_flag__msse2 - Success
-- Performing Test check_cxx_compiler_flag__msse3
-- Performing Test check_cxx_compiler_flag__msse3 - Success
-- Looking for C++ include pmmintrin.h
-- Looking for C++ include pmmintrin.h - found
-- Performing Test check_cxx_compiler_flag__mssse3
-- Performing Test check_cxx_compiler_flag__mssse3 - Success
-- Looking for C++ include tmmintrin.h
-- Looking for C++ include tmmintrin.h - found
-- Performing Test check_cxx_compiler_flag__msse4_1
-- Performing Test check_cxx_compiler_flag__msse4_1 - Success
-- Looking for C++ include smmintrin.h
-- Looking for C++ include smmintrin.h - found
-- Performing Test check_cxx_compiler_flag__msse4_2
-- Performing Test check_cxx_compiler_flag__msse4_2 - Success
-- Performing Test check_cxx_compiler_flag__mavx
-- Performing Test check_cxx_compiler_flag__mavx - Success
-- Looking for C++ include immintrin.h
-- Looking for C++ include immintrin.h - found
-- Performing Test check_cxx_compiler_flag__mfma
-- Performing Test check_cxx_compiler_flag__mfma - Success
-- Performing Test check_cxx_compiler_flag__mbmi2
-- Performing Test check_cxx_compiler_flag__mbmi2 - Success
-- Performing Test check_cxx_compiler_flag__mavx2
-- Performing Test check_cxx_compiler_flag__mavx2 - Success
-- Performing Test check_cxx_compiler_flag__mno_sse4a
-- Performing Test check_cxx_compiler_flag__mno_sse4a - Success
-- Performing Test check_cxx_compiler_flag__mno_xop
-- Performing Test check_cxx_compiler_flag__mno_xop - Success
-- Performing Test check_cxx_compiler_flag__mno_fma4
-- Performing Test check_cxx_compiler_flag__mno_fma4 - Success
-- Performing Test check_cxx_compiler_flag__mno_avx512f
-- Performing Test check_cxx_compiler_flag__mno_avx512f - Success
-- Performing Test check_cxx_compiler_flag__mno_avx512vl
-- Performing Test check_cxx_compiler_flag__mno_avx512vl - Success
-- Performing Test check_cxx_compiler_flag__mno_avx512pf
-- Performing Test check_cxx_compiler_flag__mno_avx512pf - Success
-- Performing Test check_cxx_compiler_flag__mno_avx512er
-- Performing Test check_cxx_compiler_flag__mno_avx512er - Success
-- Performing Test check_cxx_compiler_flag__mno_avx512cd
-- Performing Test check_cxx_compiler_flag__mno_avx512cd - Success
-- Performing Test check_cxx_compiler_flag__mno_avx512dq
-- Performing Test check_cxx_compiler_flag__mno_avx512dq - Success
-- Performing Test check_cxx_compiler_flag__mno_avx512bw
-- Performing Test check_cxx_compiler_flag__mno_avx512bw - Success
-- Performing Test check_cxx_compiler_flag__mno_avx512ifma
-- Performing Test check_cxx_compiler_flag__mno_avx512ifma - Success
-- Performing Test check_cxx_compiler_flag__mno_avx512vbmi
-- Performing Test check_cxx_compiler_flag__mno_avx512vbmi - Success
-- Could NOT find jemalloc (missing:  JEMALLOC_INCLUDE_DIR JEMALLOC_LIBRARIES)
-- Looking for poll.h
-- Looking for poll.h - not found
-- Looking for port.h
-- Looking for port.h - not found
-- Looking for sys/epoll.h
-- Looking for sys/epoll.h - not found
-- Looking for sys/event.h
-- Looking for sys/event.h - not found
-- Looking for sys/inotify.h
-- Looking for sys/inotify.h - not found
-- Looking for sys/select.h
-- Looking for sys/select.h - not found
-- Looking for sys/signalfd.h
-- Looking for sys/signalfd.h - not found
-- Looking for inotify_init
-- Looking for inotify_init - not found
-- Looking for epoll_ctl
-- Looking for epoll_ctl - not found
-- Looking for kqueue
-- Looking for kqueue - not found
-- Looking for port_create
-- Looking for port_create - not found
-- Looking for poll
-- Looking for poll - not found
-- Looking for select
-- Looking for select - not found
-- Looking for eventfd
-- Looking for eventfd - not found
-- Looking for signalfd
-- Looking for signalfd - not found
-- Looking for clock_gettime
-- Looking for clock_gettime - not found
-- Looking for nanosleep
-- Looking for nanosleep - not found
-- Looking for floor
-- Looking for floor - not found
-- Check if the system is big endian
-- Searching 16 bit integer
-- Looking for sys/types.h
-- Looking for sys/types.h - not found
-- Looking for stdint.h
-- Looking for stdint.h - not found
-- Looking for stddef.h
-- Looking for stddef.h - not found
-- Check size of unsigned short
-- Check size of unsigned short - failed
-- Check size of unsigned int
-- Check size of unsigned int - failed
-- Check size of unsigned long
-- Check size of unsigned long - failed
CMake Error at /usr/share/cmake-3.6/Modules/TestBigEndian.cmake:51 (message):
  no suitable type found
Call Stack (most recent call first):
  3rdParty/V8/CMakeLists.txt:26 (TEST_BIG_ENDIAN)


-- Configuring incomplete, errors occurred!
See also "arangodb-build/build/CMakeFiles/CMakeOutput.log".
See also "arangodb-build/build/CMakeFiles/CMakeError.log".
arangodb-build/build$```

With default OptimizeForArchitecture.cmake skylake is treated as merom.
``````

**Comment by dothebart** [comment]
sounds more like an issue with cmake itself being not up to date here?

**Comment by baslr** [comment]
hm could be kind of size detection issue 
```
-- Check size of unsigned short - failed
-- Check size of unsigned int
-- Check size of unsigned int - failed
-- Check size of unsigned long
-- Check size of unsigned long - failed
```

**Comment by baslr** [comment]
The architecture detection is done by a custom cmake script. This has to be updated to support the latest intel CPUs.

**Comment by Simran-B** [comment]
Will it actually optimize anything?

**Comment by baslr** [comment]
Yes, it will add compiling flags specific for the CPU.

```
-- Detected CPU: skylake
-- Performing Test check_cxx_compiler_flag__march_skylake
-- Performing Test check_cxx_compiler_flag__march_skylake - Success
-- Performing Test check_cxx_compiler_flag__msse2
-- Performing Test check_cxx_compiler_flag__msse2 - Success
-- Performing Test check_cxx_compiler_flag__msse3
-- Performing Test check_cxx_compiler_flag__msse3 - Success
... and so on.
```

**Comment by jsteemann** [comment]
@baslr @dothebart : is this issue resolved now and can be closed?

**Comment by kvahed** [comment]
This is resolved.

**Comment by baslr** [comment]
Hm, cmake with devel@acb187851f48493dec587076f7beb9133a40dcb7 on Intel Skylake

```
CMake Warning at cmake/OptimizeForArchitecture.cmake:120 (message):
  Your CPU (family 6, model 94) is not known.  Auto-detection of optimization
  flags failed and will use the 65nm Core 2 CPU settings.
Call Stack (most recent call first):
  cmake/OptimizeForArchitecture.cmake:169 (AutodetectHostArchitecture)
  CMakeLists.txt:420 (OptimizeForArchitecture)


-- Detected CPU: merom
-- Performing Test check_cxx_compiler_flag__march_merom
-- Performing Test check_cxx_compiler_flag__march_merom - Failed
-- Performing Test check_cxx_compiler_flag__march_core2
-- Performing Test check_cxx_compiler_flag__march_core2 - Success
```

**Comment by Simran-B** [comment]
Certain recent CPU models are definitely missing, most notably Skylake: https://github.com/arangodb/arangodb/blob/devel/cmake/OptimizeForArchitecture.cmake#L72

I removed my old build dir and re-generated all build files using current devel under Windows 10, but for some reason the output did not contain anything regarding the CPU detection. Am I overlooking some cache, that is **outside** the build dir?

**Comment by baslr** [comment]
```
pwd
~/arangodb-build/build$
rm -rf *
```
should be enough under GNU / Linux. But windows `¯\_(ツ)_/¯`

**Comment by kvahed** [comment]
Fixed in 2ed6d8c
